Duplicati/Library/Backend/OneDrive/MicrosoftGraphBackend.cs

```diff
@@ -796,6 +943,100 @@ private T ParseResponse<T>(HttpWebResponse response)
             }
         }
 
+        private async Task ThrowUploadSessionException(
+            UploadSession uploadSession,
+            HttpResponseMessage createSessionResponse,
+            int fragment,
+            int fragmentCount,
+            Exception ex,
+            CancellationToken cancelToken)
+        {
+            // Before throwing the exception, cancel the upload session
+            // The uploaded delete request will error if it is authenticated
+            using (HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Delete, uploadSession.UploadUrl))
+            using (HttpResponseMessage response = await this.m_client.SendAsync(request, false, cancelToken).ConfigureAwait(false))
+            {
+                // Note that the response body should always be empty in this case.
+                this.ParseResponse<UploadSession>(response);
+            }
+
+            throw new UploadSessionException(createSessionResponse, fragment, fragmentCount, ex);
+        }
+
+        private async Task ThrowUploadSessionException(
+            UploadSession uploadSession,
+            HttpWebResponse createSessionResponse,
+            int fragment,
+            int fragmentCount,
+            Exception ex,
+            CancellationToken cancelToken)
+        {
+            // Before throwing the exception, cancel the upload session
+            // The uploaded delete request will error if it is authenticated
+            var request = new AsyncHttpRequest(this.m_oAuthHelper.CreateRequest(uploadSession.UploadUrl, HttpMethod.Delete.ToString(), true));
+            using (var response = await this.m_oAuthHelper.GetResponseWithoutExceptionAsync(request, cancelToken).ConfigureAwait(false))
+            {
+                // Note that the response body should always be empty in this case.
+                this.ParseResponse<UploadSession>(response);
+            }
+
+            throw new UploadSessionException(createSessionResponse, fragment, fragmentCount, ex);
+        }
+
+        private void SetRetryAfter(RetryConditionHeaderValue retryAfter)
+        {
+            if (retryAfter != null)
+            {
+                DateTimeOffset? delayUntil = null;
+                if (retryAfter.Delta.HasValue)
+                {
+                    delayUntil = DateTimeOffset.UtcNow + retryAfter.Delta.Value;
+                }
+                else if (retryAfter.Date.HasValue)
+                {
+                    delayUntil = retryAfter.Date.Value;
+                }
+
+                if (delayUntil.HasValue)
+                {
+                    // Set the retry timestamp to the UTC version of the timestamp.
+                    Interlocked.Exchange(ref this.retryAfter, delayUntil.Value.UtcTicks);

```

Comment:
Is it possible for `SetRetryAfter` to be called twice (from two different threads) before the next call to `WaitForRetryAfter`?  In that case, would one session be possibly waiting for the incorrect amount of time?

Reply:
That is true. I've update this to make sure the newly set time is always later than the previous time - this way, in the worst case some thread may end up waiting slightly too long.